R/fonts.R

```diff
@@ -0,0 +1,442 @@
+    collapse <- if (prop %in% c("src", "unicode_range")) ", " else " "
+    x[[prop]] <- paste0(x[[prop]], collapse = collapse)
+  }
+  x$css <- font_face_css(x)
+
+  structure(x, class = "bs_font_face")
+}
+
+font_face_css <- function(x) {
+  props <- names(x)
+  font_prop <- !props %in% c("src", "unicode_range")
+  props[font_prop] <- paste0("font-", props[font_prop])
+  paste0(
+    "@font-face {\n",
+    paste0("  ", props, ": ", x, ";", collapse = "\n"),
+    "\n}"
+  )
+}
+
+#' @rdname font_face
+#' @param href A URL resource pointing to the font data.
+#' @export
+font_link <- function(family, href) {
+  structure(
+    list(family = unquote_font_family(family), href = href),
+    class = "bs_font_link"
+  )
+}
+
+#' @rdname font_face
+#' @param local Whether or not download and bundle local (woff) font files.
+#' @param cache A [sass::sass_file_cache()] object (or, more generally, a file
+#'   caching class with `$get_file()` and `$set_file()` methods). Set this
+#'   argument to `FALSE` or `NULL` to disable caching.
+#' @param wght One of the following:
+#'   * `NULL`, the default weight for the `family`.
+#'   * A character string defining an [axis range](https://developers.google.com/fonts/docs/css2#axis_ranges)
+#'   * A numeric vector of desired font weight(s).
+#' @param ital One of the following:
+#'   * `NULL`, the default `font-style` for the `family`.
+#'   * `0`, meaning `font-style: normal`
+#'   * `1`, meaning `font-style: italic`
+#'   * `c(0, 1)`, meaning both `normal` and `italic`
+#' @param display the `font-display` `@font-face` property.
+#' @references <https://developers.google.com/fonts/docs/css2>
+#' @export
+font_google <- function(family, local = TRUE, cache = sass::sass_file_cache(dir = cache_context_dir(), max_size = 100 * 1024^2),
+                        wght = NULL, ital = NULL, display = c("swap", "auto", "block", "fallback", "optional")) {
+  stopifnot(is.logical(local))
+  if (!is.null(wght)) {
+    stopifnot(is.character(wght) || is.numeric(wght))
+    wght <- sort(wght)
+  }
+  if (!is.null(ital)) {
+    stopifnot(all(ital %in% c(0, 1)))
+    ital <- sort(ital)
+  }
+  display <- match.arg(display)
+
+  axis_rng <-
+    if (is.null(wght) && is.null(ital)) {
+     ""
+    } else if (is.null(ital)) {
+      paste0(":wght@", paste0(wght, collapse = ";"))
+    } else if (is.null(wght)) {
+      paste0(":ital@", paste0(ital, collapse = ";"))
+    } else {
+      paste0(":ital,wght@", paste0(
+        apply(expand.grid(wght, ital)[, 2:1], 1, paste0, collapse = ","),
+        collapse = ";"
+      ))
+    }
+
+  structure(
+    dropNulls(list(
+      family = unquote_font_family(family),
+      local = local, cache = cache,
+      href = paste0(
+        "https://fonts.googleapis.com/css2?family=",
+        family, axis_rng, "&display=", display
+      )
+    )),
+    class = "bs_font_google"
+  )
+}
+
+is_font_object <- function(x) {
+  is_font_face(x) || is_font_link(x) || is_font_google(x)
```

Comment:
What about just having a second class for all these objects, named `bs_font`?

Reply:
Yea, that seems like a good idea

Reply:
I do need to be able to differentiate them though